### PR TITLE
Add SonarCloud Analysis to CICD (in addition to Codecov)

### PR DESCRIPTION
**Pull request by ichekrygin**
### Change 
Add Sonar[Cloud]

### Overview
Currently, `crossplane` is using [codecov](https://codecov.io/) SaaS for code coverage analysis and reporting.
This service is free for open source projects and requires minimal integration with CICD.
image

This PR replaces `codecov` with [sonarcloud](https://sonarcloud.io/about), also a SaaS code analysis service. Similar to `codecov` it is free for open source projects, and also provides code coverage report. 
Unlike `codecov`, `sonarcloud` provide additional code analysis reports: bugs, vulnerabilities, code smells, code duplication. 

image

SonarQube is integrated with CICD via Jenkins SonarScan plugin and invoked from the Jenksins pipeline. The invocation call is more detailed in terms of arguments (different set per PR vs Master).
In addition, SonarCloud provides additional (and very details) set of customization per project at the cloud interface.

### Run-time
SonarScan analysis is very quick, with duration 10-17 seconds, including uploading the report to the cloud.
```
15:36:59  INFO: ------------------------------------------------------------------------
15:36:59  INFO: EXECUTION SUCCESS
15:36:59  INFO: ------------------------------------------------------------------------
15:36:59  INFO: Total time: 16.883s
15:36:59  INFO: Final Memory: 42M/951M
```
Whereas, codecov runtime is typically 90 seconds+
```
15:37:14   / ____|        | |
15:37:14  | |     ___   __| | ___  ___ _____   __
15:37:14  | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
15:37:14  | |___| (_) | (_| |  __/ (_| (_) \ V /
15:37:14   \_____\___/ \__,_|\___|\___\___/ \_/
15:37:14                                Bash-8a28df4
15:37:14  
15:37:14  
15:37:14  [0;90m==>[0m Jenkins CI detected.
15:37:14      [0;90mproject root:[0m .
15:37:14  [0;90m-->[0m token set from env
15:37:14      [0;32mYaml not found, that's ok! Learn more at[0m [0;36mhttp://docs.codecov.io/docs/codecov-yaml[0m
15:37:14      [0;90m->[0m Found 1 reports
15:37:14  [0;90m==>[0m Detecting git/mercurial file structure
15:37:14  [0;90m==>[0m Reading reports
15:37:14      [0;32m+[0m _output/tests/linux_amd64/coverage.txt [0;90mbytes=293686[0m
15:37:14  [0;90m==>[0m Appending adjustments
15:37:14      [0;36mhttp://docs.codecov.io/docs/fixing-reports[0m
15:39:06      [0;32m+[0m Found adjustments
15:39:06  [0;90m==>[0m Gzipping contents
```

### Notifications
SonarCloud provides minimalist PR decorations in terms of PR check:
image
With additional details when expanded:
image
With further link to SonarCloud report (under SonarCloud account)

`codecov`  provide PR comment notification style with chart and changes summary
image
With further link to Codecov report (under Codecov account)

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -25,9 +25,8 @@ pipeline {
         stage('Prepare') {
             steps {
                 script {
-                    if (env.BRANCH_NAME =~ /^PR-\d+$/) {
-                        def pr_number = sh (script: "echo ${env.BRANCH_NAME} | grep -o -E '[0-9]+' ", returnStdout: true)
-                        def json = sh (script: "curl -s https://api.github.com/repos/crossplaneio/crossplane/pulls/${pr_number}", returnStdout: true).trim()
+                    if (env.CHANGE_ID == null) {
```
Comment (jbw976) [inline Jenkinsfile]: i would have expected this to be `env.CHANGE_ID != null`, meaning that we have a change_id value and can therefore use it in the curl call on the next line.  what am i missing?

Comment (ichekrygin) [reply Jenkinsfile]: That is correct, good catch.

**Comment by jbw976** [comment]
Thanks for driving the code coverage initiative @ichekrygin!!

**Comment by ichekrygin** [comment]
@jbw976 I updated this PR: (bug fix) and restored Codecov. It appears that after our changes to JenkinsCICD PR pull strategy, `codecov` numbers have improved as well. For now, I suggest we keep both  so we can keep an eye on `codecov` and also have more informed opinion on either provider.

**Pull request merged**